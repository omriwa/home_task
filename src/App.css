.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}

/*progress bar*/

.progress-bar {
    -webkit-box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
    -moz-box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
    box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
    background-color: cornflowerblue;
    height: 10px;
    transition: 750ms width;
}

/*form*/

.form {
    border: 0.5px solid gray;
    -webkit-box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
    -moz-box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
    box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
    padding: 2%;
    margin: 1% 25%;
    transition: 1s opacity;
}

.form * {
    display: block;
    margin: 0 auto;
}

/*form animation*/
.entering {
    opacity: 0;
}

.entered {
    opacity: 1;
}

.exiting {
    opacity: 1;
}

.exited {
    opacity: 0;
}