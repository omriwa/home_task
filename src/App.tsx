import * as React from 'react';
import './App.css';
// components
import ProgressBar from "./components/progressBar";
import Form from "./components/form";


interface IAppState {
    progress: number,
    fullName: string,
    phone: string,
    email: string,
    salary: string
}

class App extends React.Component<any, IAppState> {
    constructor(props: any) {
        super(props);

        this.state = {
            progress: 0,
            fullName: "",
            phone: "",
            email: "",
            salary: ""
        }

        this.renderPersonalForm = this.renderPersonalForm.bind(this);
        this.renderContantForm = this.renderContantForm.bind(this);
        this.renderSalaryForm = this.renderSalaryForm.bind(this);
        this.renderSummeryForm = this.renderSummeryForm.bind(this);
        this.setProgress = this.setProgress.bind(this);
        this.onInputChange = this.onInputChange.bind(this);
        this.setFormProgress = this.setFormProgress.bind(this);
    }

    private onInputChange(e: any): void {
        let newState = { ...this.state };
        //set the new value
        newState[e.target.name] = e.target.value;
        //update state
        this.setState(newState);
    }

    private setProgress(progress: number) {
        this.setState({
            ...this.state,
            progress: progress
        });
    }

    private setFormProgress(condition: boolean, progress: number): void {
        if (condition)
            this.setProgress(progress);
    }

    private renderPersonalForm(): any {
        return <React.Fragment>
            <label>Full Name</label>
            <input
                required
                type="text"
                value={this.state.fullName}
                name="fullName"
                onChange={this.onInputChange}
            />
        </React.Fragment>
    }

    private renderContantForm(): any {
        return <React.Fragment>
            <label>Phone Number</label>
            <input
                required
                type="tel"
                value={this.state.phone}
                name="phone"
                onChange={this.onInputChange}
            />
            <label>Email</label>
            <input
                required
                type="email"
                value={this.state.email}
                name="email"
                onChange={this.onInputChange}
            />
        </React.Fragment>
    }

    private validateEmail(email: string): boolean {
        let pattern = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return pattern.test(String(email).toLowerCase());
    }

    private renderSalaryForm(): any {
        return <React.Fragment>
            <h4>Salary</h4>
            <label>0-1000</label>
            <input
                onChange={this.onInputChange}
                value={"0-1000"}
                type="radio"
                name="salary"
            />
            <label>1000-2000</label>
            <input
                onChange={this.onInputChange}
                value={"1000-2000"}
                type="radio"
                name="salary"
            />
            <label>2000-3000</label>
            <input
                onChange={this.onInputChange}
                value={"´2000-3000"}
                type="radio"
                name="salary"
            />
            <label>3000-4000</label>
            <input
                onChange={this.onInputChange}
                value={"3000-4000"}
                type="radio"
                name="salary"
            />
            <label>4000-more</label>
            <input
                onChange={this.onInputChange}
                value={"4000-more"}
                type="radio"
                name="salary"
            />
        </React.Fragment>
    }

    private renderSummeryForm(): any {
        let stateCopy = { ...this.state },
            keys;
        //remove progress
        delete stateCopy.progress;
        //get all state keys
        keys = (Object as any).keys(stateCopy);

        return <React.Fragment>
            <h4>Summery</h4>
            {
                keys.map(key => {
                    return <div>{stateCopy[key]}</div>
                })
            }

        </React.Fragment>
    }

    public render() {
        return (
            <div className="App">
                <ProgressBar progress={this.state.progress} />

                <Form
                    in={this.state.progress === 0}
                    setProgress={() => this.setFormProgress(this.state.fullName !== "", 30)}
                >
                    {
                        this.renderPersonalForm()
                    }
                </Form>

                <Form
                    in={this.state.progress === 30}
                    setProgress={() => this.setFormProgress(this.state.phone.length === 10 && this.validateEmail(this.state.email), 60)}
                >
                    {
                        this.renderContantForm()
                    }
                </Form>

                <Form
                    in={this.state.progress === 60}
                    setProgress={() => this.setFormProgress(this.state.salary !== "", 90)}
                >
                    {
                        this.renderSalaryForm()
                    }
                </Form>

                <Form
                    in={this.state.progress === 90}
                    setProgress={() => this.setFormProgress(true, 100)}
                >
                    {
                        this.renderSummeryForm()
                    }
                </Form>

            </div>
        );
    }
}

export default App;
