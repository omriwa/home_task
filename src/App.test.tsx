import * as React from 'react';
import App from './App';
import ProgressBar from "./components/progressBar";

import * as renderer from 'react-test-renderer';

test("app render", () => {
    const app = renderer.create(<App />);
    //check render
    let tree = app.toJSON();
    expect(tree).toMatchSnapshot();
});

test("progress bar", () => {
    const progressBar = renderer.create(<ProgressBar progress={0} />);
    //check render
    let tree = progressBar.toJSON();
    expect(tree).toMatchSnapshot();
});

