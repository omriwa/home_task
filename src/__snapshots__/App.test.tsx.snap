// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app filling progress 1`] = `
<div
  className="App"
>
  <div>
    <div
      className="progress"
    >
      <div
        aria-valuemax={100}
        aria-valuemin="0"
        aria-valuenow={0}
        className="progress-bar"
        role="progressbar"
        style={
          Object {
            "width": "0%",
          }
        }
      />
    </div>
  </div>
  <div
    className="form entered"
  >
    <label>
      Full Name
    </label>
    <input
      name="fullName"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <div>
      <button
        onClick={[Function]}
      >
        next
      </button>
    </div>
  </div>
</div>
`;

exports[`app render 1`] = `
<div
  className="App"
>
  <div>
    <div
      className="progress"
    >
      <div
        aria-valuemax={100}
        aria-valuemin="0"
        aria-valuenow={0}
        className="progress-bar"
        role="progressbar"
        style={
          Object {
            "width": "0%",
          }
        }
      />
    </div>
  </div>
  <div
    className="form entered"
  >
    <label>
      Full Name
    </label>
    <input
      name="fullName"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <div>
      <button
        onClick={[Function]}
      >
        next
      </button>
    </div>
  </div>
</div>
`;

exports[`progress bar 1`] = `
<div>
  <div
    className="progress"
  >
    <div
      aria-valuemax={100}
      aria-valuemin="0"
      aria-valuenow={0}
      className="progress-bar"
      role="progressbar"
      style={
        Object {
          "width": "0%",
        }
      }
    />
  </div>
</div>
`;
